### ISSUE TITLE
Deploying custom object detection model on stm32n6 nucleo board- camera doesn't detect

### ISSUE THREAD

**Issue by faiqahmadkhan** [STMicroelectronics/stm32ai-modelzoo-services#46]
Hi, I am trying to deploy a custom object detection model on the STM32N6-Nucleo board. 
I started with this:
https://github.com/STMicroelectronics/stm32ai-modelzoo-services/tree/main/object_detection
The training gets done, I get the performance metrics,
python stm32ai_main.py
 The deployment code works fine, and I get the deployment complete message. But when I change from development mode to flash boot mode. It doesn't detect the camera. 

When I program the demo camera.hex file
https://github.com/STMicroelectronics/x-cube-n6-camera-capture/tree/main/Binary/NUCLEO-N657X0-Q
 or object detection .hex file, 
https://github.com/STMicroelectronics/STM32N6-GettingStarted-ObjectDetection/tree/main/Binary/NUCLEO-N657X0-Q/USB-UVC-Display
the camera works perfectly.

Below are the training and deployment configuration files. These are the only files I changed. I would really appreciate any help in this regard. Thank you, 

# training configure.yaml file:
general:
  project_name: custom_food_yoloxn_256
  logs_dir: logs
  saved_models_dir: saved_models
  display_figures: false
  gpu_memory_limit: 16 
  num_threads_tflite: 12
  global_seed: 127

operation_mode: chain_tqe

model:
  model_type: st_yoloxn
  model_name: st_yoloxn_d033_w025
  input_shape: (256,256,3)

dataset:
  format: darknet_yolo
  dataset_name: darknet_yolo
  class_names: [beverage, food]
  exclude_unlabeled: true
  download_data: false
  max_detections: 50
  train_images_path: path/yolov5/yolo_dataset_prep/yolo_ready_dataset_416/images/train
  train_annotations_path: path/r/yolov5/yolo_dataset_prep/yolo_ready_dataset_416/labels/train
  val_images_path: path/yolov5/yolo_dataset_prep/yolo_ready_dataset_416/images/val
  val_annotations_path: path/yolov5/yolo_dataset_prep/yolo_ready_dataset_416/labels/val
  test_images_path: path/yolov5/yolo_dataset_prep/yolo_ready_dataset_416/images/test
  test_annotations_path: path/yolov5/yolo_dataset_prep/yolo_ready_dataset_416/labels/test
  test_path: path/yolov5/yolo_dataset_prep/yolo_ready_dataset_416/images/tfs_labels/test
  quantization_path: path/yolov5/yolo_dataset_prep/yolo_ready_dataset_416/images/test

preprocessing:
  rescaling:
    scale: 1/255
    offset: 0
  resizing:
    aspect_ratio: fit
    interpolation: nearest
  color_mode: rgb

data_augmentation:
  random_contrast:
    factor: 0.4
  random_brightness:
    factor: 0.3
  random_flip:
    mode: horizontal
  random_translation:
    width_factor: 0.15
    height_factor: 0.15
    fill_mode: reflect
    interpolation: nearest
  random_rotation:
    factor: 0.02
    fill_mode: reflect
    interpolation: nearest
  random_crop:
    crop_center_x: (0.25, 0.75)
    crop_center_y: (0.25, 0.75)
    crop_width: (0.6, 0.9)
    crop_height: (0.6, 0.9)
    change_rate: 0.9

training:
  dropout: null
  batch_size: 64
  epochs: 100
  optimizer:
    Adam:
      learning_rate: 0.0025
  callbacks:
    LRWarmupCosineDecay:
      initial_lr: 1.0e-05
      warmup_steps: 20
      max_lr: 0.00125
      hold_steps: 20
      decay_steps: 300
      end_lr: 1.0e-06
    EarlyStopping:
      monitor: val_loss
      patience: 60
      restore_best_weights: true
      verbose: 1
quantization:
  quantizer: TFLite_converter
  quantization_type: PTQ
  quantization_input_type: uint8
  quantization_output_type: int8
  export_dir: quantized_models
postprocessing:
  confidence_thresh: 0.300
  NMS_thresh: 0.5
  IoU_eval_thresh: 0.5
  plot_metrics: true
  max_detection_boxes: 100

mlflow:
   uri: ./tf/src/experiments_outputs/mlruns

hydra:
   run:
      dir: ./tf/src/experiments_outputs/custom_food_yoloxn_256

# deploying custom model configure file
general:
  project_name: custom_food_yoloxn_256
  logs_dir: logs
  saved_models_dir: saved_models
  display_figures: false
  gpu_memory_limit: 16 #24
  num_threads_tflite: 12
  global_seed: 127

operation_mode: deployment

model:
  model_type: st_yoloxn
  model_path: C:/stm32zoo/stm32ai-modelzoo-services/object_detection/tf/src/experiments_outputs/custom_food_yoloxn_256/quantized_models/quantized_model.tflite

dataset:
  format: darknet_yolo
  dataset_name: darknet_yolo
  class_names: [beverage, food]
  exclude_unlabeled: true
  download_data: false
  max_detections: 50
  train_images_path: path/yolov5/yolo_dataset_prep/yolo_ready_dataset_416/images/train
  train_annotations_path: path/r/yolov5/yolo_dataset_prep/yolo_ready_dataset_416/labels/train
  val_images_path: path/yolov5/yolo_dataset_prep/yolo_ready_dataset_416/images/val
  val_annotations_path: path/yolov5/yolo_dataset_prep/yolo_ready_dataset_416/labels/val
  test_images_path: path/yolov5/yolo_dataset_prep/yolo_ready_dataset_416/images/test
  test_annotations_path: path/yolov5/yolo_dataset_prep/yolo_ready_dataset_416/labels/test
  test_path: path/yolov5/yolo_dataset_prep/yolo_ready_dataset_416/images/tfs_labels/test
  quantization_path: path/yolov5/yolo_dataset_prep/yolo_ready_dataset_416/images/test

preprocessing:
  rescaling:
    scale: 1/255
    offset: 0
  resizing:
    aspect_ratio: fit
    interpolation: nearest
  color_mode: rgb

data_augmentation:
  random_contrast:
    factor: 0.4
  random_brightness:
    factor: 0.3
  random_flip:
    mode: horizontal
  random_translation:
    width_factor: 0.15
    height_factor: 0.15
    fill_mode: reflect
    interpolation: nearest
  random_rotation:
    factor: 0.02
    fill_mode: reflect
    interpolation: nearest
  random_crop:
    crop_center_x: (0.25, 0.75)
    crop_center_y: (0.25, 0.75)
    crop_width: (0.6, 0.9)
    crop_height: (0.6, 0.9)
    change_rate: 0.9

training:
  dropout: null
  batch_size: 64
  epochs: 100
  optimizer:
    Adam:
      learning_rate: 0.0025
  callbacks:
    LRWarmupCosineDecay:
      initial_lr: 1.0e-05
      warmup_steps: 20
      max_lr: 0.00125
      hold_steps: 20
      decay_steps: 300
      end_lr: 1.0e-06
    EarlyStopping:
      monitor: val_loss
      patience: 60
      restore_best_weights: true
      verbose: 1
quantization:
  quantizer: TFLite_converter
  quantization_type: PTQ
  quantization_input_type: uint8
  quantization_output_type: int8
  export_dir: quantized_models
postprocessing:
  confidence_thresh: 0.300
  NMS_thresh: 0.5
  IoU_eval_thresh: 0.5
  plot_metrics: true
  max_detection_boxes: 10

tools:
  stedgeai:
    optimization: balanced
    on_cloud: False
    path_to_stedgeai: C:/ST/STEdgeAI/4.0/Utilities/windows/stedgeai.exe
  path_to_cubeIDE: C:/ST/STM32CubeIDE_2.0.0/STM32CubeIDE/stm32cubeide.exe

deployment:
  c_project_path: ../application_code/object_detection/STM32N6/
  IDE: GCC
  verbosity: 1
  hardware_setup:
    serie: STM32N6
    board: NUCLEO-N657X0-Q
    output: "UVCL"

mlflow:
   uri: ./tf/src/experiments_outputs/mlruns

hydra:
   run:
      dir: ./tf/src/experiments_outputs/custom_food_yoloxn_256

**Comment by GRATTINSTM** [comment]
Hello @faiqahmadkhan,

Deploying the default `st_yoloxn_d033_w025_256_int8.tflite`, I don't have any problem getting the camera to work. Can you try to remove the "training" and "data_augmentation" sections from the deployment YAML file? Make sure to power cycle the board when the flash is done.

**Comment by GRATTINSTM** [comment]
@faiqahmadkhan could you share your `quantized_model.tflite`?

**Comment by faiqahmadkhan** [comment]
@GRATTINSTM thank you so much for helping out. I will try the deployment again by removing the training and data augmentation sections and will update you. Here is the quantized .tflite file attached. Could you please test it? I would really appreciate that. Thanks,
link to model: https://alabama.box.com/s/12uoqd901ms8gf8kx3d8g4xa2xy00x86

**Comment by faiqahmadkhan** [comment]
@GRATTINSTM I also tried quantized_tiny_yolo_v2_224_.tflite, but the camera didn't work. 
the log and configure content:

# the log:
(st_env) C:\stm32zoo\stm32ai-modelzoo-services\object_detection>python stm32ai_main.py
[INFO] : Running `deployment` operation mode
[INFO] : Using provided class names from dataset.class_names
cfg.yolo_anchors: [0.5, 0.5]
[INFO] : ClearML config check
[INFO] : Setting upper limit of usable GPU memory to 16GBytes.
[INFO] : The random seed for this simulation is 127
Loading model from C:/stm32zoo/stm32ai-modelzoo-services/application_code/object_detection/STM32N6/Model/quantized_tiny_yolo_v2_224_.tflite
INFO: Created TensorFlow Lite XNNPACK delegate for CPU.
[INFO] : Generating C header file for Getting Started...
application_code/audio/STM32N6
application_code/face_detection/STM32N6
application_code/hand_posture/STM32F4
application_code/image_classification/STM32N6
application_code/image_classification/STM32H7
application_code/instance_segmentation/STM32N6
application_code/object_detection/STM32N6
[WARNING]: Submodule 'application_code/object_detection/STM32N6' has uncommitted changes. Please commit or stash them.
loading model.. model_path="C:/stm32zoo/stm32ai-modelzoo-services/application_code/object_detection/STM32N6/Model/quantized_tiny_yolo_v2_224_.tflite"
loading conf file.. "../application_code/object_detection/STM32N6/stmaic_NUCLEO-N657X0-Q.conf" config="UVCL"
"UVCL_ModelZoo" configuration is used
[INFO] : Selected board :  "NUCLEO-N657X0-Q Getting Started Object Detection (STM32CubeIDE)" (stm32_cube_ide/UVCL_ModelZoo/stm32n6)
[INFO] : Compiling the model and generating optimized C code + Lib/Inc files:  C:/stm32zoo/stm32ai-modelzoo-services/application_code/object_detection/STM32N6/Model/quantized_tiny_yolo_v2_224_.tflite
setting STM.AI tools.. root_dir="", req_version=""
 Cube AI Path: "C:\ST\STEdgeAI\4.0\Utilities\windows\stedgeai.exe".
[INFO] : Offline CubeAI used; Selected tools:  12.0.0 (x-cube-ai pack)
loading conf file.. "../application_code/object_detection/STM32N6/stmaic_NUCLEO-N657X0-Q.conf" config="UVCL"
"UVCL_ModelZoo" configuration is used
compiling... "quantized_tiny_yolo_v2_224__tflite" session
 model_path  : C:/stm32zoo/stm32ai-modelzoo-services/application_code/object_detection/STM32N6/Model/quantized_tiny_yolo_v2_224_.tflite
 tools       : 12.0.0 (x-cube-ai pack)
 target      : "NUCLEO-N657X0-Q Getting Started Object Detection (STM32CubeIDE)" (stm32_cube_ide/UVCL_ModelZoo/stm32n6)
 options     : --st-neural-art default@../application_code/object_detection/STM32N6/Model/user_neuralart_NUCLEO-N657X0-Q.json --input-data-type uint8 --inputs-ch-position chlast --output-data-type int8
"series" value is not coherent.. stm32n6 != stm32n6npu
 results -> RAM=351,232 IO=0:0 WEIGHTS=11,053,488 MACC=0 RT_RAM=13 RT_FLASH=56,847 LATENCY=0.000
[INFO] : Optimized C code + Lib/Inc files generation done.
[INFO] : Building the STM32 c-project..
deploying the c-project.. "NUCLEO-N657X0-Q Getting Started Object Detection (STM32CubeIDE)" (stm32_cube_ide/UVCL_ModelZoo/stm32n6)
updating.. UVCL_ModelZoo
 -> s:copying file.. "network.c" to ..\application_code\object_detection\STM32N6\Model\NUCLEO-N657X0-Q\network.c
 -> s:copying file.. "network_ecblobs.h" to ..\application_code\object_detection\STM32N6\Model\NUCLEO-N657X0-Q\network_ecblobs.h
 -> u:copying file.. "stai_network.c" to ..\application_code\object_detection\STM32N6\Model\NUCLEO-N657X0-Q\stai_network.c
 -> u:copying file.. "stai_network.h" to ..\application_code\object_detection\STM32N6\Model\NUCLEO-N657X0-Q\stai_network.h
 -> u:copying file.. "network_atonbuf.xSPI2.raw" to ..\application_code\object_detection\STM32N6\Model\NUCLEO-N657X0-Q\network_atonbuf.xSPI2.raw
 -> u:copying file.. "app_config.h" to ..\application_code\object_detection\STM32N6\Application\NUCLEO-N657X0-Q\Inc\app_config.h
 -> searching for pre-existing stedgeai-lib from "C:\stm32zoo\stm32ai-modelzoo-services\application_code\object_detection\STM32N6\Application\NUCLEO-N657X0-Q\STM32CubeIDE"
 -> updating stedgeai-lib at "C:\stm32zoo\stm32ai-modelzoo-services\application_code\object_detection\STM32N6\Middlewares\stedgeai-lib" from "C:\stm32zoo\stm32ai-modelzoo-services\object_detection\tf\src\experiments_outputs\custom_food_yoloxn_256\generated\stedgeai-lib"
 -> updating cproject file with "NetworkRuntime1200_CM55_GCC.a"
building.. UVCL_ModelZoo
flashing.. UVCL_ModelZoo NUCLEO-N657X0-Q
[INFO] deployment complete

(st_env) C:\stm32zoo\stm32ai-modelzoo-services\object_detection>

# deployment config file:
general:
  project_name: custom_food_yoloxn_256
  logs_dir: logs
  saved_models_dir: saved_models
  display_figures: false
  gpu_memory_limit: 16 #24
  num_threads_tflite: 12
  global_seed: 127


operation_mode: deployment

model:
  model_type: st_yoloxn

  model_path: C:/stm32zoo/stm32ai-modelzoo-services/application_code/object_detection/STM32N6/Model/quantized_tiny_yolo_v2_224_.tflite

dataset:
  format: darknet_yolo
  dataset_name: darknet_yolo
  class_names: [beverage, food]
  exclude_unlabeled: true
  download_data: false
  max_detections: 50
train_images_path: path/yolov5/yolo_dataset_prep/yolo_ready_dataset_416/images/train
train_annotations_path: path/r/yolov5/yolo_dataset_prep/yolo_ready_dataset_416/labels/train
val_images_path: path/yolov5/yolo_dataset_prep/yolo_ready_dataset_416/images/val
val_annotations_path: path/yolov5/yolo_dataset_prep/yolo_ready_dataset_416/labels/val
test_images_path: path/yolov5/yolo_dataset_prep/yolo_ready_dataset_416/images/test
test_annotations_path: path/yolov5/yolo_dataset_prep/yolo_ready_dataset_416/labels/test
test_path: path/yolov5/yolo_dataset_prep/yolo_ready_dataset_416/images/tfs_labels/test
quantization_path: path/yolov5/yolo_dataset_prep/yolo_ready_dataset_416/images/test

preprocessing:
  rescaling:
    scale: 1/255
    offset: 0
  resizing:
    aspect_ratio: fit
    interpolation: nearest
  color_mode: rgb


quantization:
  quantizer: TFLite_converter
  quantization_type: PTQ
  quantization_input_type: uint8
  quantization_output_type: int8
  export_dir: quantized_models
postprocessing:
  confidence_thresh: 0.300
  NMS_thresh: 0.5
  IoU_eval_thresh: 0.5
  plot_metrics: true
  max_detection_boxes: 10


tools:
  stedgeai:
    optimization: balanced
    on_cloud: False
    path_to_stedgeai: C:/ST/STEdgeAI/4.0/Utilities/windows/stedgeai.exe

  path_to_cubeIDE: C:/ST/STM32CubeIDE_2.0.0/STM32CubeIDE/stm32cubeide.exe

deployment:
  c_project_path: ../application_code/object_detection/STM32N6/
  IDE: GCC
  verbosity: 1
  hardware_setup:
    serie: STM32N6
    board: NUCLEO-N657X0-Q
    output: "UVCL"

mlflow:
   uri: ./tf/src/experiments_outputs/mlruns

hydra:
   run:
      dir: ./tf/src/experiments_outputs/custom_food_yoloxn_256

**Comment by GRATTINSTM** [comment]
Everything is working fine on my side. Find attached some images I took with the Windows Camera application, and the `user_config.yaml` file I used.

<img width="320" height="240" alt="Image" src="https://github.com/user-attachments/assets/a9f32715-6a8d-4e11-b19b-740a6601454f" />


<img width="320" height="240" alt="Image" src="https://github.com/user-attachments/assets/1526f5d3-bf0d-4cf6-aca3-08cb942c36e1" />


[user_config.yaml](https://github.com/user-attachments/files/27156781/user_config.yaml)

**Comment by faiqahmadkhan** [comment]
Hi @GRATTINSTM thank you so much. I think the issue is with the way I do a power cycle or cables. Maybe this works when I upload the demo .hex file through Cubeprogrammer, but not for development through the script. 
This is how I do it:
1. connect cable to CN10
2. connect to development mode 
3. run the deployment code, 
when it is completed
4. change to boot flash mode
5. disconnect from cn10
after few seconds
6. connect to CN10 and CN8

In this way when i used to upload .hex file from https://github.com/STMicroelectronics/x-cube-n6-camera-capture/tree/main/Binary/NUCLEO-N657X0-Q , through the Cube Programmer it would work. 

<img width="4032" height="3024" alt="Image" src="https://github.com/user-attachments/assets/1e429aa9-e61f-43eb-81fa-2a86847d2a5b" />
<img width="5712" height="4284" alt="Image" src="https://github.com/user-attachments/assets/256c8636-640f-46cc-8410-17e87e3b1dab" />

**Comment by GRATTINSTM** [comment]
Can you try to first set the development mode and then power cycle the board before running the deployment script?

**Comment by faiqahmadkhan** [comment]
@GRATTINSTM  yes, I tried that, but it didn't work out.

**Comment by GRATTINSTM** [comment]
It's weird because the ModelZoo programs the board the exact same way you program it with STM32CubeProgrammer, there's no reason why it should work in one case and not in the other... Plus the ModelZoo checks the programming commands' output so any fail would result in an error in the script...

**Comment by faiqahmadkhan** [comment]
@GRATTINSTM  for object detection also CubeIDE v1.17.0 works. I was using CubeIDE v2.0.0. It resolved. Thank you so much.
